### ISSUE TITLE
Hi and thanks for your efforts. Is there an example for how todo this?

### ISSUE THREAD

**Issue by huysentruitw** [huysentruitw/pem-utils#24]
Hi and thanks for your efforts. Is there an example for how todo this?
I have case where I am creating a `DerAsnIa5String` and serializes perfectly!
```csharp
var contextSpesificId = new DerAsnIdentifier(DerAsnTagClass.ContextSpecific, DerAsnEncodingType.Primitive, DerAsnKnownTypeTags.Primitive.ObjectIdentifier);
var location = new DerAsnIa5String(contextSpesificId, uri.ToString());
// etc....
```
But cant seem to get the value out. I tried to hack my way through by doing `Encoding.ASCII.GetString(objectIdentifier.Value.Select(x => (byte)x).ToArray())` which gives me almost the string I wanted with missing one character plus two control characters at the begining. Did not like this approach anyway.

What i am trying to figure out is how to tell the default decoder not to parse this into `DerAsnObjectIdentifier` if it sees the `DerAsnTagClass.ContextSpecific` but instead parse it into `DerAsnIa5String`. Can it be done?

_Originally posted by @cleftheris in https://github.com/huysentruitw/pem-utils/issues/16#issuecomment-483612370_

**Comment by huysentruitw** [comment]
@cleftheris can you please specify a unit-test case that explains what you're trying to do?

**Comment by cleftheris** [comment]
Hi @huysentruitw,

I have had success with the library both in encoding and decoding structures for `X509Certificate2` extensions missing from dotnet. Here is [the code](https://github.com/indice-co/Indice.Psd2/tree/master/src/Indice.Psd2.Cryptography/X509Certificates) I am using right now and I managed to create custom DerAsnTypes and register them with the encoder/decoder per use case. 
- [ContextSpecificSequence.cs](https://github.com/indice-co/Indice.Psd2/blob/master/src/Indice.Psd2.Cryptography/X509Certificates/DerAsnTypes/ContextSpecificSequence.cs)	
- [ContextSpecificString.cs](https://github.com/indice-co/Indice.Psd2/blob/master/src/Indice.Psd2.Cryptography/X509Certificates/DerAsnTypes/ContextSpecificString.cs)
- [DerAsnEnumerated.cs](https://github.com/indice-co/Indice.Psd2/blob/master/src/Indice.Psd2.Cryptography/X509Certificates/DerAsnTypes/DerAsnEnumerated.cs)	
- [OctetStringSequence.cs](https://github.com/indice-co/Indice.Psd2/blob/master/src/Indice.Psd2.Cryptography/X509Certificates/DerAsnTypes/OctetStringSequence.cs)

These work magnificently on upon encoding although I had a diffcult time decoding the `DerAsnEnumerated` and `OctetStringSequence` because these are competing with the basic structures in regards to identification. I dont know if it is even possible but it would be great to conditionaly pick the class type for decoding based on something more. What I did to workaround this for example in the case of a OctetStringEncoded sequence of Enumerated values [was simply to skip some bytes ahead](https://github.com/indice-co/Indice.Psd2/blob/3433028a9183bd679badd3b9035286a0455b7b44/src/Indice.Psd2.Cryptography/X509Certificates/CertificateRevocationList.cs#L187). The same with other cases

Summing up there is no issue I could not workaround and I am pretty overwelmed be the ease of use for the DerConverter. I have not figured out the difference between registering generic types and types but if there is something I am doing wrong that I could make better please tell me so.

I am posting here for reference and feedback so feel free to close the issue. 

Cheers

**Comment by huysentruitw** [comment]
I'm not sure why bytes need to be skipped, I didn't find so much information about the subject. Anyway, the difference between `RegisterType` and `RegisterGenericType` is that, for the latter, the class is ignored. So registering a type with `RegisterGenericType` will be used in the decoding process regardless of the `DerAsnTagClass`. When registering a type with `RegisterType`, the registered type will only be used by the decoder when the class, encodingType and tag match.

If you feel like any of those types should be in the DerConverter package, please let me know. But make sure this is part of the DER specification and not BER. While BER can be supported by registering your own types, it isn't the purpose of this library.

Regards.
